pkg/controller/dbsyncer/generic_db_to_transport_syncer.go

```diff
@@ -81,28 +76,70 @@ func (syncer *genericDBToTransportSyncer) initLastUpdateTimestampFromTransport()
 }
 
 func (syncer *genericDBToTransportSyncer) syncBundle(ctx context.Context) {
-	lastUpdateTimestamp, err := syncer.db.GetLastUpdateTimestamp(ctx, syncer.dbTableName)
-	if err != nil {
-		syncer.log.Error(err, "unable to sync bundle to leaf hubs", syncer.dbTableName)
-		return
-	}
+	ticker := time.NewTicker(syncer.intervalPolicy.GetInterval())
 
-	if !lastUpdateTimestamp.After(*syncer.lastUpdateTimestamp) { // sync only if something has changed
-		return
+	for {
+		select {
+		case <-ctx.Done(): // we have received a signal to stop
+			ticker.Stop()
+			return
+
+		case <-ticker.C:
+			lastUpdateTimestamp, err := syncer.db.GetLastUpdateTimestamp(ctx, syncer.dbTableName)
+			if err != nil {
+				syncer.completeSync(ticker, false)
+				syncer.log.Error(err, "unable to sync bundle to leaf hubs", syncer.dbTableName)
+
+				continue
+			}
+
+			if !lastUpdateTimestamp.After(*syncer.lastUpdateTimestamp) { // sync only if something has changed
+				syncer.completeSync(ticker, false)
+
+				continue
+			}
+
+			// if we got here, then the last update timestamp from db is after what we have in memory.
+			// this means something has changed in db, syncing to transport.
+			bundleResult := syncer.createBundleFunc()
+			lastUpdateTimestamp, err = syncer.db.GetBundle(ctx, syncer.dbTableName, syncer.createObjFunc, bundleResult)
+
+			if err != nil {
+				syncer.completeSync(ticker, false)
+				syncer.log.Error(err, "unable to sync bundle to leaf hubs", syncer.dbTableName)
+
+				continue
+			}
+
+			syncer.lastUpdateTimestamp = lastUpdateTimestamp
+
+			syncer.syncToTransport(syncer.transportBundleKey, datatypes.SpecBundle, lastUpdateTimestamp, bundleResult)
+			syncer.completeSync(ticker, true)
+		}
 	}
-	// if we got here, then the last update timestamp from db is after what we have in memory.
-	// this means something has changed in db, syncing to transport.
-	bundleResult := syncer.createBundleFunc()
-	lastUpdateTimestamp, err = syncer.db.GetBundle(ctx, syncer.dbTableName, syncer.createObjFunc, bundleResult)
+}
 
-	if err != nil {
-		syncer.log.Error(err, "unable to sync bundle to leaf hubs", syncer.dbTableName)
-		return
+// completeSync notifies policy whether sync was actually performed or skipped and resets ticker's interval
+// to a new recalculated one.
+func (syncer *genericDBToTransportSyncer) completeSync(ticker *time.Ticker, syncPerformed bool) {
+	// get current sync interval
+	currentInterval := syncer.intervalPolicy.GetInterval()
+
+	// notify policy whether sync was actually performed or skipped
+	if syncPerformed {
+		syncer.intervalPolicy.OnSyncPerformed()
+	} else {
+		syncer.intervalPolicy.OnSyncSkipped()
 	}
 
-	syncer.lastUpdateTimestamp = lastUpdateTimestamp
+	// get recalculated sync interval
+	recalculatedInterval := syncer.intervalPolicy.GetInterval()
 
-	syncer.syncToTransport(syncer.transportBundleKey, datatypes.SpecBundle, lastUpdateTimestamp, bundleResult)
+	ticker.Reset(recalculatedInterval)
```

Comment:
why do you reset the ticker in cases nothing has changed?
anyway you're checking if two lines later so it could be inside the if..

Reply:
done